### PR TITLE
Document the debug window.

### PR DESCRIPTION
**Pull request by andersio**
The document describes the Debug Window in its ideal state. Some essential parts are not implemented yet, and are marked with a 🚧 emoji.

### CODE REVIEW

File: Cookbook/Technical-Documents/TheDebugWindow.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+Sometimes there is a complicated user journey, requiring multiple steps or data entry to proceed. Sometimes there is a prerequisite to reach a certain user flow that is awfully long for recurring manual QA and dev testing.
+
+You want to have an early way out to this prolonged, endured process of having to repeat yourselves just to get through them.
+
+## 😘 The Debug Window is here to help! 🎉
+
+## What is the Debug Window?
+
+It always stands by and wait for you call — either
```
Comment (RuiAAPeres) [inline Cookbook/Technical-Documents/TheDebugWindow.md]: I don't understand this sentence.

File: Cookbook/Technical-Documents/TheDebugWindow.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
+Integrate with the Debug Window
+==========================================
+
+Sometimes there is a complicated user journey, requiring multiple steps or data entry to proceed. Sometimes there is a prerequisite to reach a certain user flow that is awfully long for recurring manual QA and dev testing.
+
+You want to have an early way out to this prolonged, endured process of having to repeat yourselves just to get through them.
```
Comment (RuiAAPeres) [inline Cookbook/Technical-Documents/TheDebugWindow.md]: Maybe a less a dramatic sentence? 😛

File: Cookbook/Technical-Documents/TheDebugWindow.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+The Debug Window serves these functionalities:
+
+* Provide a list of contextual debug actions.
+* Quick access to useful debug information, like the current environment and the app version.
+
+## What are debug actions?
+
+Represented in code as `DebugAction`, these are debug shortcuts or side effects that are supported by your screen.
+
+All `DebugAction` has a title, which need not be localized. You may configure them to be either of these three forms:
```
Comment (RuiAAPeres) [inline Cookbook/Technical-Documents/TheDebugWindow.md]: `All `DebugAction` has a title` ➡️ `All `DebugAction` have a title`

**Comment by Ben-Henshall** [comment]
I currently have a PR open that affects the documentation for this tool, so please do not merge until that branch has been merged (#7228), so that I can update the documentation.

File: Cookbook/Technical-Documents/TheDebugWindow.md (hunk 2 above)
Comment (andersio) [reply Cookbook/Technical-Documents/TheDebugWindow.md]: Why

File: Cookbook/Technical-Documents/TheDebugWindow.md (hunk 1 above)
Comment (Ben-Henshall) [reply Cookbook/Technical-Documents/TheDebugWindow.md]:
```suggestion
The debug window can be activated in the following ways:
```

File: Cookbook/Technical-Documents/TheDebugWindow.md (hunk 3 above)
Comment (Ben-Henshall) [reply Cookbook/Technical-Documents/TheDebugWindow.md]:
```suggestion
All `DebugAction` have a title, which need not be localized. You may configure them to be either of these three forms:
```

File: Cookbook/Technical-Documents/TheDebugWindow.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+Sometimes there is a complicated user journey, requiring multiple steps or data entry to proceed. Sometimes there is a prerequisite to reach a certain user flow that is awfully long for recurring manual QA and dev testing.
+
+You want to have an early way out to this prolonged, endured process of having to repeat yourselves just to get through them.
+
+## 😘 The Debug Window is here to help! 🎉
+
+## What is the Debug Window?
+
+It always stands by and waits for you call — either
```
Comment (Ben-Henshall) [inline Cookbook/Technical-Documents/TheDebugWindow.md]:
```suggestion
The debug window can be activated in the following ways:
```

File: Cookbook/Technical-Documents/TheDebugWindow.md (hunk 1 above)
Comment (Ben-Henshall) [reply Cookbook/Technical-Documents/TheDebugWindow.md]: fixed

**Review by NSMyself — APPROVED** [review]
Nice write-up 👍

**Review by inamiy — CHANGES_REQUESTED** [review]
This document is an ideal version which I think this needs to pass the proposal first.

I still don't get what will be the differences from current state.
(There should be more 🚧 printed in current 8d87703 documentation)

**Comment by inamiy** [comment]
Or, we can review the actual implementation first, then this documentation.

**Comment by Ben-Henshall** [comment]
@inamiy I can remove the 🚧  parts for now, and ask for feedback on other possible improvements.
I'm not sure if this warrants a proposal though, considering they are minor improvements to an existing system?

**Comment by andersio** [comment]
@inamiy I asked @RuiAAPeres and the conclusion was that there is no proposal for this, since both debug functionalities covered by the document already exist for a fairly long time. The ideal state is just unifying the implementations of both.

**Comment by andersio** [comment]
For @inamiy and others stumbled upon this document, as noted in the document, there are only three parts not implemented yet:

1. `FormsViewController` forwarding `DebugActionsProviding` calls to its View Model.
2. Extend the Debug Window infrastructure to support shake, and move all shake-to-fill actions to be under the same umbrella.
3. "Share Quick Info" button.

**Comment by inamiy** [comment]
My current understanding is:

- "Shake" and "debug menu" is different
- This proposal is trying to make "shake" as shortcut for opening "debug menu"
- There is no "ubiquitous action" yet

so current state and ideal state are quite far apart now... Am I right?

I think the best flow would be: 1. proposal -> 2. implementation -> 3. documentation.
If we skip proposal, we want to see real implementation before merging this documentation.
Just let me know, am I taking a right step on this? (or too verbose?)

**Comment by andersio** [comment]
I don't mind taking out the "to-be-done" bits as a separate branch if that's desired.

> There is no "ubiquitous action" yet
This bit is in already.

File: Cookbook/Technical-Documents/TheDebugWindow.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+## What are debug actions?
+
+Represented in code as `DebugAction`, these are debug shortcuts or side effects that are supported by your screen.
+
+All `DebugAction` have a title, which need not be localized. You may configure them to be either of these three forms:
+
+* Simple action — a simple one-off closure invocation;
+* Producer action — starting a `SignalProducer`, and showing an alert upon failure; and
+* Modal action — a factory closure which generates a `UIViewController`, which would be presented by the Debug Window.
+* Date action - A one-off closure that retrieves a selected (optional) date. You could also provide an initially selected date.
```
Comment (inamiy) [inline Cookbook/Technical-Documents/TheDebugWindow.md]: Is this related to `DebugAction` ? 
It seems it relates to "debug window" alongside "gender picker" though.

Comment (Ben-Henshall) [reply Cookbook/Technical-Documents/TheDebugWindow.md]: Sort of.

The date picker has been in there for a while (component showcase) is separate from any debug actions. I'm not sure if we need to keep it in there any longer to be honest.

A new `DebugAction` was added a few days ago that included this date action, which looks the same as the component showcase version, but can be used by screens to set certain date-related flags.

Comment (inamiy) [reply Cookbook/Technical-Documents/TheDebugWindow.md]: I'm only seeing the type `DebugAction` and `DebugAction.Action` as a proof-of-concept, so I think we should not include information unrelated to these types.

Comment (Ben-Henshall) [reply Cookbook/Technical-Documents/TheDebugWindow.md]: Sorry, I should clarify. `date` *is* a `DebugAction.Action` case. By "sort of", I meant that it looks identical to the component showcase picker (which has 0 function)

Comment (inamiy) [reply Cookbook/Technical-Documents/TheDebugWindow.md]: OK, maybe we can move date-picker as another `DebugAction` with some special flag to place at the bottom? 🤔 

I still want to see it appearing in real code somehow, but for now, I think I understand the similar concept, so I will move on 👍

**Review by inamiy — APPROVED** [review]

File: Cookbook/Technical-Documents/TheDebugWindow.md (hunk 5 above)
Comment (Ben-Henshall) [reply Cookbook/Technical-Documents/TheDebugWindow.md]: This is already in code: #7228 (in Babylon repo)

Comment (inamiy) [reply Cookbook/Technical-Documents/TheDebugWindow.md]: Oh I was looking at wrong branch... #7228 solves my issue 👍

**Pull request merged**